commit c07a6e8cdc2c2c0222d2441863df534b96377f35
Author: Nicholas Wilson <nicholas@nicholaswilson.me.uk>
Date:   Tue Dec 2 16:21:32 2025 +0000

    Fix regression in handling of repeated backslash escapes in pcre2_substitute (#857)

diff --git a/src/pcre2_substitute.c b/src/pcre2_substitute.c
index fb187258..f323c56a 100644
--- a/src/pcre2_substitute.c
+++ b/src/pcre2_substitute.c
@@ -115,6 +115,7 @@ for (; ptr < ptrend; ptr++)
     int erc;
     int errorcode;
     uint32_t ch;
+    PCRE2_SPTR esc_end_ptr;
 
     if (ptr < ptrend - 1) switch (ptr[1])
       {
@@ -137,6 +138,9 @@ for (; ptr < ptrend; ptr++)
       goto EXIT;
       }
 
+    esc_end_ptr = ptr;
+    ptr -= 1;  /* Rewind by one, because the for-loop will increment it */
+
     switch(erc)
       {
       case 0:      /* Data character */
@@ -160,7 +164,8 @@ for (; ptr < ptrend; ptr++)
 
       default:
       if (erc < 0)
-          break;  /* capture group reference */
+        break;  /* capture group reference */
+      ptr = esc_end_ptr;
       rc = PCRE2_ERROR_BADREPESCAPE;
       goto EXIT;
       }
diff --git a/testdata/testinput2 b/testdata/testinput2
index 8a905f58..e0cacced 100644
--- a/testdata/testinput2
+++ b/testdata/testinput2
@@ -8314,5 +8314,12 @@ a)"xI
     foo|bar\=offset=1,substitute_subject=F|OOBAR,copy_matched_subject
     foo|bar\=offset=1,substitute_subject=F|OOBAR,copy_matched_subject,zero_terminate
 
+# --------------
+# Some more tests for a substitution regression
+# --------------
+
+/foo(?<Bar>BAR)?/substitute_extended,replace=X${Bar:+\:\:text}Y
+    foo
+    fooBAR
 
 # End of testinput2
diff --git a/testdata/testoutput2 b/testdata/testoutput2
index 08e73823..bdd431e3 100644
--- a/testdata/testoutput2
+++ b/testdata/testoutput2
@@ -23656,6 +23656,15 @@ Failed: error -72: substitute subject differs from prior match call
     foo|bar\=offset=1,substitute_subject=F|OOBAR,copy_matched_subject,zero_terminate
 Failed: error -72: substitute subject differs from prior match call
 
+# --------------
+# Some more tests for a substitution regression
+# --------------
+
+/foo(?<Bar>BAR)?/substitute_extended,replace=X${Bar:+\:\:text}Y
+    foo
+ 1: XY
+    fooBAR
+ 1: X::textY
 
 # End of testinput2
 Error -80: PCRE2_ERROR_BADDATA (unknown error number)
